crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,800 @@
+    if opt_a2 != Some(true) {
+        for ctrl_index in 0..num_qubits {
+            let [um00, um11, um01, um10] = extract_multiplex_blocks(mat, ctrl_index);
+
+            if off_diagonals_are_zero(&um01, &um10, None) {
+                return Ok(demultiplex(
+                    &um00,
+                    &um11,
+                    opt_a1_val,
+                    false, // opt_a2
+                    depth,
+                    Some(num_qubits - 1 - ctrl_index),
+                    VWType::All,
+                )?
+                .0);
+            }
+        }
+    }
+    let opt_a2_val = opt_a2.unwrap_or(true);
+    let out_qubits = (0..num_qubits)
+        .map(|_| ShareableQubit::new_anonymous())
+        .collect::<Vec<_>>();
+    let mut out = CircuitData::new(Some(out_qubits), None, None, 0, Param::Float(0.)).unwrap();
+    // perform block ZXZ decomposition from [2]
+    let (a1, a2, b, c) = _block_zxz_decomp(mat);
+
+    debug_assert!(_zxz_decomp_verify(mat, &a1, &a2, &b, &c));
+
+    let iden = DMatrix::<Complex64>::identity(dim / 2, dim / 2);
+    let (left_circuit, vmat_c, _) = demultiplex(
+        &iden,
+        &c,
+        opt_a1_val,
+        opt_a2_val,
+        depth,
+        None,
+        VWType::OnlyW,
+    )?;
+    let (right_circuit, _, wmat_a) =
+        demultiplex(&a1, &a2, opt_a1_val, opt_a2_val, depth, None, VWType::OnlyV)?;
+
+    // middle circ
+    // zmat is needed in order to reduce two cz gates, and combine them into the B2 matrix
+    let mut zmat = DMatrix::<Complex64>::zeros(dim / 2, dim / 2);
+    for i in 0..dim / 2 {
+        zmat[(i, i)] = if i < dim / 4 {
+            Complex64::from(1.0)
+        } else {
+            Complex64::from(-1.0)
+        };
+    }
+    // wmatA and vmatC are combined into B1 and B2
+    let b1 = &wmat_a * &vmat_c;
+    let b2 = if opt_a1_val {
+        &zmat * &wmat_a * &b * &vmat_c * &zmat
+    } else {
+        &wmat_a * &b * &vmat_c
+    };
+    let middle_circ = demultiplex(&b1, &b2, opt_a1_val, opt_a2_val, depth, None, VWType::All)?.0;
+
+    // the output circuit of the block ZXZ decomposition from [2]
+    let qr = (0..num_qubits).map(Qubit::new).collect::<Vec<_>>();
+    append(&mut out, left_circuit, &qr)?;
+    let _ = out.push_standard_gate(StandardGate::H, &[], &[Qubit((num_qubits - 1) as u32)]);
+    append(&mut out, middle_circ, &qr)?;
+    let _ = out.push_standard_gate(StandardGate::H, &[], &[Qubit((num_qubits - 1) as u32)]);
+    append(&mut out, right_circuit, &qr)?;
+    if opt_a2_val && depth == 0 && dim > 4 {
+        apply_a2(&out, two_qubit_decomposer)
+    } else {
+        Ok(out)
+    }
+}
+
+fn zxz_decomp_svd(a: DMatrixView<Complex64>) -> (DMatrix<Complex64>, DMatrix<Complex64>) {
+    let (v, sigma, w_dg) = svd_decomposition(a);
+
+    let s = &v * &sigma * &v.adjoint();
+    let u = v * w_dg;
+
+    (s, u)
+}
+
+/// Block ZXZ decomposition method, by Krol and Al-Ars [2]
+fn _block_zxz_decomp(
```

Comment:
We don't need the leading underscore in rust the function is already private.

```suggestion
fn block_zxz_decomp(
```